cmd/kubeadm/app/checks/checks.go

```diff
@@ -0,0 +1,185 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package checks
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"os/exec"
+)
+
+// PreFlightCheck validates the state of the system to ensure kubeadm will be
+// successful as often as possilble.
+type PreFlightCheck interface {
+	Check() (warnings []string, errors []string)
+}
+
+// ServiceCheck verifies that the given service is enabled and active. If we do not
+// detect a supported init system however, all checks are skipped and a warning is
+// returned.
+type ServiceCheck struct {
+	service string
+}
+
+func (sc ServiceCheck) Check() (warnings []string, errors []string) {
+
+	initSystem := getInitSystem()
+	if initSystem == nil {
+		return []string{"no supported init system detected, skipping service checks"}, nil
+	}
+
+	warnings = []string{}
+
+	if !initSystem.ServiceExists(sc.service) {
+		warnings = append(warnings, fmt.Sprintf("%s service does not exist", sc.service))
+		return warnings, nil
+	}
+
+	if !initSystem.ServiceIsEnabled(sc.service) {
+		errors = append(errors,
+			fmt.Sprintf("%s service is not enabled, please run 'systemctl enable %s.service'",
+				sc.service, sc.service))
+	}
+
+	if !initSystem.ServiceIsActive(sc.service) {
+		errors = append(errors,
+			fmt.Sprintf("%s service is not active, please run 'systemctl start %s.service'",
+				sc.service, sc.service))
+	}
+
+	return warnings, nil
+}
+
+// PortOpenCheck ensures the given port is available for use.
+type PortOpenCheck struct {
+	port int
+}
+
+func (poc PortOpenCheck) Check() (warnings []string, errors []string) {
+	errors = []string{}
+	// TODO: Get IP from KubeadmConfig
+	ln, err := net.Listen("tcp", fmt.Sprintf(":%d", poc.port))
+	if err != nil {
+		errors = append(errors, fmt.Sprintf("Port %d is in use.", poc.port))
+	}
+	if ln != nil {
+		ln.Close()
+	}
+
+	return nil, errors
+}
+
+// DirAvailableCheck checks if the given directory either does not exist, or
+// is empty.
+type DirAvailableCheck struct {
+	path string
+}
+
+func (dac DirAvailableCheck) Check() (warnings []string, errors []string) {
+	errors = []string{}
+	// If it doesn't exist we are good:
+	if _, err := os.Stat(dac.path); os.IsNotExist(err) {
+		return nil, nil
+	}
+
+	f, err := os.Open(dac.path)
+	if err != nil {
+		errors = append(errors, fmt.Sprintf("Unable to check if %s is empty: %s", dac.path, err))
+		return nil, errors
+	}
+	defer f.Close()
+
+	_, err = f.Readdirnames(1)
+	if err != io.EOF {
+		errors = append(errors, fmt.Sprintf("%s is not empty", dac.path))
+	}
+
+	return nil, errors
+}
+
+// InPathChecks checks if the given executable is present in the path.
+type InPathCheck struct {
+	executable string
```

Comment:
You should add a field here that's indicating if the binary is mandatory.
If `mandatory = true` then an error should be thrown, otherwise not